File: Week 9/stack.c
#include <stdio.h>
#include <stdlib.h>

#define MAX 100  // Maksimum kapasitas stack

// Struktur stack
typedef struct {
    int top;
    int array[MAX];
} Stack;

// Inisialisasi stack
void init(Stack *s) {
    s->top = -1;
}

// Periksa apakah stack kosong
int isEmpty(Stack *s) {
    return s->top == -1;
}

// Periksa apakah stack penuh
int isFull(Stack *s) {
    return s->top == MAX - 1;
}

// Push elemen ke stack
void push(Stack *s, int data) {
    if (isFull(s)) {
        printf("Stack overflow!\n");
        return;
    }
    s->array[++(s->top)] = data;
}

// Pop elemen dari stack
int pop(Stack *s) {
    if (isEmpty(s)) {
        printf("Stack underflow!\n");
        return -1;
    }
    return s->array[(s->top)--];
}

// Mengintip elemen teratas stack
int peek(Stack *s) {
    if (isEmpty(s)) {
        printf("Stack kosong!\n");
        return -1;
    }
    return s->array[s->top];
}

// Menampilkan seluruh elemen dalam stack
void show(Stack *s) {
    if (isEmpty(s)) {
        printf("Stack kosong!\n");
        return;
    }
    printf("Stack: ");
    for (int i = 0; i <= s->top; i++) {
        printf("%d ", s->array[i]);
    }
    printf("\n");
}

int main() {
    Stack s;
    init(&s);

    push(&s, 10);
    push(&s, 20);
    push(&s, 30);

    printf("Sebelum pop:\n");
    show(&s);
    pop(&s);
    printf("Setelah pop:\n");
    show(&s);
    return 0;
}

File: Week 9/stack.py
class Stack: 
    def __init__(self):
        self.stack = []

    def push(self, data):
            self.stack.append(data)
    
    def pop(self):
         if not self.isEmpty():
            return self.stack.pop()
         else:
            return None

    #check apa kosong atau
        
    #return data kalau dia tidak kosong
    def peek(self):
        if not self.isEmpty():
            return self.stack[-1]
        else:
            return None    
    def isEmpty(self):
        return len(self.stack) == 0    
    
    def show(self):
        return self.stack
  
s = Stack()
s.push(10)
s.push(20)
s.push(30)

print(s.show())

print(s.pop())

print(s.show())